File: frontend/driver/app/src/main/java/com/mju/onda/driver/feature/settings/data/SafeStopApi.kt
package com.mju.onda.driver.feature.settings.data

import android.util.Log
import com.mju.onda.driver.core.supabase.SupabaseClient
import java.net.URLEncoder
import java.time.Instant
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

/**
 * Supabase `safe_stop_requests` — 안전 정차 요청 전송·조회·취소.
 * (이전 Vite `/api/safe-stop` 의존 제거)
 */
object SafeStopApi {
    private const val TAG = "SafeStopApi"

    data class RemoteRequest(
        val id: String,
        val decision: String,
        val reason: String,
        val requestedAt: String,
        val routeName: String,
        val vehicleName: String,
        val date: String,
    )

    sealed class PostResult {
        data class Ok(val id: String) : PostResult()
        data object Failed : PostResult()
    }

    sealed class FetchResult {
        data class Ok(val items: List<RemoteRequest>) : FetchResult()
        data object Failed : FetchResult()
    }

    suspend fun postRequest(
        id: String,
        driverId: String,
        driverName: String,
        vehicleName: String,
        routeName: String,
        operationId: String,
        reason: String,
        detailReason: String,
        requestedAt: String,
        date: String,
    ): PostResult = withContext(Dispatchers.IO) {
        if (!SupabaseClient.isConfigured || SupabaseClient.accessToken.isNullOrBlank()) {
            Log.w(TAG, "Supabase not ready")
            return@withContext PostResult.Failed
        }
        val driverUuid = SupabaseClient.userUuid
        if (driverUuid.isNullOrBlank()) {
            Log.w(TAG, "No user uuid")
            return@withContext PostResult.Failed
        }

        try {
            val opUuid = resolveOperationUuid(operationId)
            val body = JSONObject().apply {
                put("driver_id", driverUuid)
                put("reason", reason)
                put("detail_reason", detailReason)
                put("decision", "pending")
                put("requested_at", Instant.now().toString())
                if (!opUuid.isNullOrBlank()) put("operation_id", opUuid)
            }

            val result = SupabaseClient.request(
                method = "POST",
                path = "/rest/v1/safe_stop_requests",
                jsonBody = body.toString(),
                authed = true,
            )
            if (result.code !in 200..299) {
                Log.w(TAG, "POST failed HTTP ${result.code}: ${result.body.take(240)}")
                return@withContext PostResult.Failed
            }
            val serverId = parseInsertedId(result.body) ?: id
            Log.d(TAG, "posted safe_stop id=$serverId op=$opUuid")
            // keep unused params referenced for call-site compatibility
            Log.d(TAG, "meta driver=$driverId/$driverName $vehicleName $routeName $requestedAt $date")
            PostResult.Ok(serverId)
        } catch (e: Exception) {
            Log.w(TAG, "POST error: ${e.message}")
            PostResult.Failed
        }
    }

    suspend fun cancelRequest(id: String): PostResult = withContext(Dispatchers.IO) {
        if (!SupabaseClient.isConfigured || SupabaseClient.accessToken.isNullOrBlank() || id.isBlank()) {
            return@withContext PostResult.Failed
        }
        try {
            val patch = JSONObject()
                .put("decision", "cancelled")
                .put("decided_at", Instant.now().toString())
            val result = SupabaseClient.request(
                method = "PATCH",
                path = "/rest/v1/safe_stop_requests?id=eq.${enc(id)}",
                jsonBody = patch.toString(),
                authed = true,
            )
            if (result.code !in 200..299) {
                Log.w(TAG, "cancel failed HTTP ${result.code}: ${result.body.take(200)}")
                return@withContext PostResult.Failed
            }
            PostResult.Ok(id)
        } catch (e: Exception) {
            Log.w(TAG, "cancel error: ${e.message}")
            PostResult.Failed
        }
    }

    /**
     * @param driverId login_id 또는 uuid — 실제 조회는 세션 uuid 우선
     */
    suspend fun fetchForDriver(driverId: String): FetchResult = withContext(Dispatchers.IO) {
        if (!SupabaseClient.isConfigured || SupabaseClient.accessToken.isNullOrBlank()) {
            return@withContext FetchResult.Failed
        }
        val uuid = SupabaseClient.userUuid
            ?: driverId.takeIf { it.contains('-') && it.length > 20 }
            ?: return@withContext FetchResult.Failed

        try {
            val select = listOf(
                "id",
                "decision",
                "reason",
                "requested_at",
                "created_at",
                "operations:operation_id(operation_date,buses:bus_id(bus_name),schedules:schedule_id(routes:route_id(route_name)))",
            ).joinToString(",")
            val path =
                "/rest/v1/safe_stop_requests?select=${enc(select)}" +
                    "&driver_id=eq.${enc(uuid)}" +
                    "&order=created_at.desc"

            val result = SupabaseClient.request(method = "GET", path = path, authed = true)
            if (result.code !in 200..299) {
                Log.w(TAG, "GET failed HTTP ${result.code}: ${result.body.take(200)}")
                return@withContext FetchResult.Failed
            }
            FetchResult.Ok(parse(result.body))
        } catch (e: Exception) {
            Log.w(TAG, "GET error: ${e.message}")
            FetchResult.Failed
        }
    }

    private fun resolveOperationUuid(operationId: String): String? {
        if (operationId.isBlank() || operationId == "unknown") return null
        if (operationId.contains('-') && operationId.length >= 32 &&
            !operationId.startsWith("week-") && !operationId.startsWith("op-") &&
            !operationId.startsWith("d02-") && !operationId.startsWith("stop-")
        ) {
            // likely uuid
            return operationId
        }
        // try as uuid anyway if matches uuid pattern
        val uuidRegex = Regex(
            "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$",
        )
        if (uuidRegex.matches(operationId)) return operationId

        val byExt = SupabaseClient.request(
            method = "GET",
            path = "/rest/v1/operations?select=id&external_id=eq.${enc(operationId)}&limit=1",
            authed = true,
        )
        if (byExt.code in 200..299) {
            val arr = JSONArray(byExt.body)
            if (arr.length() > 0) return arr.getJSONObject(0).optString("id").ifBlank { null }
        }
        val byId = SupabaseClient.request(
            method = "GET",
            path = "/rest/v1/operations?select=id&id=eq.${enc(operationId)}&limit=1",
            authed = true,
        )
        if (byId.code in 200..299) {
            val arr = JSONArray(byId.body)
            if (arr.length() > 0) return arr.getJSONObject(0).optString("id").ifBlank { null }
        }
        Log.w(TAG, "operation not resolved: $operationId")
        return null
    }

    private fun parseInsertedId(body: String): String? {
        val trimmed = body.trim()
        if (trimmed.startsWith("[")) {
            val arr = JSONArray(trimmed)
            if (arr.length() == 0) return null
            return arr.getJSONObject(0).optString("id").ifBlank { null }
        }
        if (trimmed.startsWith("{")) {
            return JSONObject(trimmed).optString("id").ifBlank { null }
        }
        return null
    }

    private fun parse(json: String): List<RemoteRequest> {
        val arr = JSONArray(json)
        val out = ArrayList<RemoteRequest>(arr.length())
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            val ops = o.optJSONObject("operations")
            val buses = ops?.optJSONObject("buses")
            val routes = ops?.optJSONObject("schedules")?.optJSONObject("routes")
            val requestedAt = o.optString("requested_at")
            out.add(
                RemoteRequest(
                    id = o.getString("id"),
                    decision = o.optString("decision", "pending"),
                    reason = o.optString("reason", ""),
                    requestedAt = toHhmm(requestedAt),
                    routeName = routes?.optString("route_name").orEmpty(),
                    vehicleName = buses?.optString("bus_name").orEmpty(),
                    date = ops?.optString("operation_date").orEmpty(),
                ),
            )
        }
        return out
    }

    private fun toHhmm(iso: String): String {
        if (iso.length >= 16 && iso[10] == 'T') {
            // 2026-08-11T07:06:17... → use local-ish HH:mm from string (UTC) ok for display
            return iso.substring(11, 16)
        }
        return iso.take(5)
    }

    private fun enc(value: String): String = URLEncoder.encode(value, "UTF-8")
}

File: frontend/admin/src/lib/safeStopApi.ts
import { isSupabaseConfigured, supabase } from './supabase'
import { todayDateKey } from '../types/assignment'
import type { Database } from '../types/database'

export type SafeStopDecision = 'pending' | 'continue' | 'stop' | 'cancelled'

export type SafeStopRequest = {
  id: string
  driverId: string
  driverName: string
  vehicleName: string
  routeName: string
  operationId: string
  reason: string
  detailReason: string
  requestedAt: string
  date: string
  decision: SafeStopDecision
  createdAt: number
  decidedAt?: number
}

type SafeStopRow = Database['public']['Tables']['safe_stop_requests']['Row'] & {
  users: { id: string; name: string; login_id: string | null } | null
  operations: {
    id: string
    external_id: string | null
    operation_date: string
    buses: { bus_name: string } | null
    schedules: { routes: { route_name: string } | null } | null
  } | null
}

const SELECT = `
  id,
  operation_id,
  driver_id,
  reason,
  detail_reason,
  decision,
  requested_at,
  decided_at,
  created_at,
  users:driver_id ( id, name, login_id ),
  operations:operation_id (
    id,
    external_id,
    operation_date,
    buses:bus_id ( bus_name ),
    schedules:schedule_id ( routes:route_id ( route_name ) )
  )
`

function hhmm(iso: string | null | undefined): string {
  if (!iso) return ''
  const d = new Date(iso)
  if (Number.isNaN(d.getTime())) {
    // already HH:mm
    return iso.slice(0, 5)
  }
  const p = (n: number) => String(n).padStart(2, '0')
  return `${p(d.getHours())}:${p(d.getMinutes())}`
}

function rowToRequest(row: SafeStopRow): SafeStopRequest {
  const op = row.operations
  return {
    id: row.id,
    driverId: row.users?.login_id || row.driver_id,
    driverName: row.users?.name || '',
    vehicleName: op?.buses?.bus_name || '미정',
    routeName: op?.schedules?.routes?.route_name || '-',
    operationId: op?.external_id || op?.id || row.operation_id || '',
    reason: row.reason,
    detailReason: row.detail_reason || '',
    requestedAt: hhmm(row.requested_at),
    date: op?.operation_date || todayDateKey(),
    decision: row.decision,
    createdAt: row.created_at ? Date.parse(row.created_at) : Date.now(),
    decidedAt: row.decided_at ? Date.parse(row.decided_at) : undefined,
  }
}

async function fetchFromDevApi(params?: {
  driverId?: string
  pendingOnly?: boolean
}): Promise<SafeStopRequest[]> {
  try {
    const qs = new URLSearchParams()
    if (params?.driverId) qs.set('driverId', params.driverId)
    if (params?.pendingOnly) qs.set('pending', '1')
    const res = await fetch(`/api/safe-stop?${qs}`, { cache: 'no-store' })
    if (!res.ok) return []
    const data = (await res.json()) as SafeStopRequest[]
    return Array.isArray(data) ? data : []
  } catch {
    return []
  }
}

export async function fetchSafeStopRequests(params?: {
  driverId?: string
  pendingOnly?: boolean
  limit?: number
}): Promise<SafeStopRequest[]> {
  if (!isSupabaseConfigured) return fetchFromDevApi(params)

  let q = supabase.from('safe_stop_requests').select(SELECT).order('created_at', { ascending: false })

  if (params?.pendingOnly) q = q.eq('decision', 'pending')
  if (params?.limit != null) q = q.limit(params.limit)
  else q = q.limit(50)
  if (params?.driverId) {
    // login_id 또는 uuid
    if (params.driverId.includes('-') && params.driverId.length > 20) {
      q = q.eq('driver_id', params.driverId)
    } else {
      const { data: user } = await supabase
        .from('users')
        .select('id')
        .eq('login_id', params.driverId)
        .maybeSingle()
      if (!user?.id) return []
      q = q.eq('driver_id', user.id)
    }
  }

  const { data, error } = await q
  if (error) {
    console.error('[safe_stop]', error.message)
    return []
  }
  return ((data ?? []) as unknown as SafeStopRow[]).map(rowToRequest)
}

export async function decideSafeStop(
  id: string,
  decision: 'continue' | 'stop' | 'cancelled',
): Promise<{ ok: boolean; message?: string }> {
  if (!isSupabaseConfigured) {
    try {
      const res = await fetch(`/api/safe-stop/${encodeURIComponent(id)}`, {
        method: 'PATCH',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ decision }),
      })
      const data = (await res.json()) as { ok?: boolean; message?: string }
      if (!res.ok) return { ok: false, message: data.message || 'failed' }
      return { ok: true }
    } catch {
      return { ok: false, message: 'network error' }
    }
  }

  const { data: current, error: findErr } = await supabase
    .from('safe_stop_requests')
    .select('id, decision, operation_id')
    .eq('id', id)
    .maybeSingle()

  if (findErr) return { ok: false, message: findErr.message }
  if (!current) return { ok: false, message: '요청을 찾을 수 없습니다' }
  if (current.decision !== 'pending') return { ok: false, message: '이미 처리된 요청입니다' }

  const { error } = await supabase
    .from('safe_stop_requests')
    .update({
      decision,
      decided_at: new Date().toISOString(),
    })
    .eq('id', id)

  if (error) return { ok: false, message: error.message }

  // 중단 승인 시 operations를 즉시 COMPLETED로 바꾸지 않음
  // → 실시간 목록에서는 계속 '안전 정차'로 보이고, 기사 측 종료 처리 시 완료로 전환
  if (decision === 'continue' && current.operation_id) {
    // 계속 운행: 운행 중 유지(이미 IN_PROGRESS면 변경 없음)
    const { error: opErr } = await supabase
      .from('operations')
      .update({
        status: 'IN_PROGRESS',
        ended_at: null,
      })
      .eq('id', current.operation_id)
    if (opErr) console.warn('[safe_stop] operation resume', opErr.message)
  }

  return { ok: true }
}
